cyten/planar.py

```diff
@@ -0,0 +1,400 @@
+            raise ValueError('Tensor must be fully labelled')
+        if tensor.num_legs != len(self.labels):
+            raise ValueError('Mismatching number of legs')
+        if not tensor.labels_are(*self.labels):
+            raise ValueError('Mismatching labels')
+        offset = tensor.get_leg_idcs(self.labels[0])
+        if tensor._labels != self.labels[-offset:] + self.labels[:-offset]:
+            raise ValueError(f'Labels must match placeholder up to cyclical permutation. '
+                             f'Expected {", ".join(self.labels)}. Got {", ".join(tensor._labels)}')
+
+    def copy(self):
+        return TensorPlaceholder(labels=self.labels[:], dims=self.dims[:])
+
+    def dagger(self):
+        return TensorPlaceholder(labels=[_dual_leg_label(l) for l in reversed(self.labels)],
+                                 dims=self.dims[::-1])
+
+    @property
+    def fully_specified(self) -> bool:
+        """If the connectivity of all legs is specified.
+
+        Fully specified means that each leg, or rather its :attr:`label` appears in either
+        the :attr:`contracted_legs` or the :attr:`open_legs`.
+        """
+        for l in self.labels:
+            if l not in self.contracted_legs and l not in self.open_legs:
+                return False
+        return True
+
+    def __getitem__(self, idx):
+        if not isinstance(idx, str):
+            msg = f'Expected string label. Got {type(idx)}'
+            raise TypeError(msg)
+        if idx not in self.labels:
+            msg = f'Invalid label: "{idx}". Expected one of {", ".join(self.labels)}'
+            raise ValueError(msg)
+        return LegPlaceholder(self, idx)
+
+
+class PlanarDiagram(metaclass=ABCMeta):
+    """TODO: mention "access points". 
+
+    Pitfalls / warnings::
+        - clockwise order of labels
+
+    """
+
+    @classmethod
+    @abstractmethod
+    def define_diagram(cls) -> list[TensorPlaceholder]:
+        """TODO"""
+        ...
+
+    @classmethod
+    def evaluate(cls, *tensors: Tensor) -> Tensor:
+        """Perform the contraction specified by the diagram.
+
+        Parameters
+        ----------
+        tensors : list of Tensor
+            The tensors in the diagram. Must
+        """
+        # Note: subclasses should probably override this with a thin wrapper that has a named
+        #       argument for each tensor, to make the order explicit.
+        tensors = list(tensors)
+        placeholders = cls.define_diagram()
+        order = cls.get_contraction_order(tensors=tensors, placeholders=placeholders)
+        return contract(tensors=tensors, placeholders=placeholders, order=order)
+
+    @classmethod
+    def get_contraction_order(cls, tensors: list[Tensor], placeholders: list[TensorPlaceholder]
+                              ) -> list[int]:
+        """Determine the order for binary contractions during :meth:`do_binary_contractions`.
+
+        Returns
+        -------
+        order : list of int
```

Comment:
Not enough, generically you need a tree for the contraction order?

Reply:
yes